eth/state_accessor.go

```diff
@@ -72,138 +82,59 @@ func (eth *Ethereum) stateAtBlock(block *types.Block, reexec uint64) (statedb *s
 		logged time.Time
 		parent common.Hash
 	)
-	defer func() {
-		if err != nil && parent != (common.Hash{}) {
-			database.TrieDB().Dereference(parent)
-		}
-	}()
-	for block.NumberU64() < origin {
+	for current.NumberU64() < origin {
 		// Print progress logs if long enough time elapsed
-		if time.Since(logged) > 8*time.Second {
-			log.Info("Regenerating historical state", "block", block.NumberU64()+1, "target", origin, "remaining", origin-block.NumberU64()-1, "elapsed", time.Since(start))
+		if time.Since(logged) > 8*time.Second && report {
+			log.Info("Regenerating historical state", "block", current.NumberU64()+1, "target", origin, "remaining", origin-current.NumberU64()-1, "elapsed", time.Since(start))
 			logged = time.Now()
 		}
 		// Retrieve the next block to regenerate and process it
-		if block = eth.blockchain.GetBlockByNumber(block.NumberU64() + 1); block == nil {
-			return nil, nil, fmt.Errorf("block #%d not found", block.NumberU64()+1)
+		if current = eth.blockchain.GetBlockByNumber(current.NumberU64() + 1); current == nil {
+			return nil, fmt.Errorf("block #%d not found", current.NumberU64()+1)
```

Comment:
You're assigning to `current` here, and if `current` is `nil`, you try to access `current.NumberU64`. This was erroneous previously too, see https://github.com/ethereum/go-ethereum/issues/22485#issuecomment-796981882

Reply:
fixed